pageserver/client_grpc/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+//!
+//! Pageserver gRPC client library
+//!
+//! This library provides a gRPC client for the pageserver for the
+//! communicator project.
+//!
+//! This library is a work in progress.
+//!
+//!
+
+use std::collections::HashMap;
+use bytes::Bytes;
+use futures::{StreamExt};
+use thiserror::Error;
+use tonic::metadata::AsciiMetadataValue;
+use pageserver_page_api::proto;
+use pageserver_page_api::proto::PageServiceClient;
+use utils::shard::ShardIndex;
+use std::fmt::Debug;
+use tracing::error;
+use tokio::sync::RwLock;
+use tonic::transport::{Channel, Endpoint};
+
+#[derive(Error, Debug)]
+pub enum PageserverClientError {
+    #[error("could not connect to service: {0}")]
+    ConnectError(#[from] tonic::transport::Error),
+    #[error("could not perform request: {0}`")]
+    RequestError(#[from] tonic::Status),
+    #[error("protocol error: {0}")]
+    ProtocolError(#[from] pageserver_page_api::ProtocolError),
+    #[error("could not perform request: {0}`")]
+    InvalidUri(#[from] http::uri::InvalidUri),
+    #[error("could not perform request: {0}`")]
+    Other(String),
+}
```

Comment:
We should think about how callers should do error handling, especially considering that they will have to handle both `GetPageStatusCode` and `tonic::Status`. We can consider adding conversion between `GetPageStatusCode` to `tonic::Status` such that callers only need to handle one kind of error. We should also think about splitting out some of these errors to construction-time (e.g. `InvalidUri`), since callers often have to do exhaustive matching on all variants and it's kind of annoying to have a bunch of variants that can never happen for typical method calls. But we can probably defer that as a TODO until we start building the higher layers.